### PR TITLE
Cancellation Tokens

### PR DESCRIPTION
**Pull request by JoeCooper**
Please have a look at how I've done this.

I've added cancellation tokens with default tokens as default parameters.

Then, I've added a cancellation token parameter to the algorithm base. I understand this is not intended as a public interface and it does not, therefore, offer a default parameter.

At each stage I've amended the unit tests to compile, however I do not have a plan for unit testing the cancellation tokens themselves.

You might notice at this point most code is not _observing_ the tokens. According to Microsoft's guidelines, a method is not required to act on the token's state, and it is specifically not advised to abort after the primary workload has completed.

With the interfaces in order, various persons can then add code to observe the tokens in strategic points.

My third commit here ("Observe cancellation token in SearchClosestEdge") evaluates the cancellation token (with .ThrowIfCancellationRequested) at the start of two loop bodies in HilbertExtensions.SearchClosestEdge. This is because these loops are where the bulk of the time is spent in the default Router's TryResolve implementation.

If I have time (unconfirmed) I would like to go find a good spot to check for cancellation in Router.TryCalculateRaw, but we should sync up first.

### CODE REVIEW

File: src/Itinero/Algorithms/Routes/CompleteRouteBuilder.cs [hunk 1]
```diff
@@ -470,36 +470,36 @@ private void AppendRoute(Route route, IAttributeCollection shortcutMeta)
         /// <summary>
         /// Builds a route.
         /// </summary>
-        public static Route Build(RouterDb db, Profile profile, RouterPoint source, RouterPoint target, EdgePath<float> path)
+        public static Route Build(RouterDb db, Profile profile, RouterPoint source, RouterPoint target, EdgePath<float> path, CancellationToken cancellationToken)
```
Comment (airbreather) [inline src/Itinero/Algorithms/Routes/CompleteRouteBuilder.cs]: Missed defaulting here?

File: src/Itinero/Algorithms/Routes/CompleteRouteBuilder.cs [hunk 2]
```diff
@@ -470,36 +470,36 @@ private void AppendRoute(Route route, IAttributeCollection shortcutMeta)
...
         {
-            return CompleteRouteBuilder.TryBuild(db, profile, source, target, path).Value;
+            return CompleteRouteBuilder.TryBuild(db, profile, source, target, path, cancellationToken).Value;
         }
 
         /// <summary>
         /// Builds a route.
         /// </summary>
-        public static Result<Route> TryBuild<T>(RouterDb db, Profile profile, RouterPoint source, RouterPoint target, EdgePath<T> path)
+        public static Result<Route> TryBuild<T>(RouterDb db, Profile profile, RouterPoint source, RouterPoint target, EdgePath<T> path, CancellationToken cancellationToken)
```
Comment (airbreather) [inline src/Itinero/Algorithms/Routes/CompleteRouteBuilder.cs]: Missed defaulting here?

File: src/Itinero/Algorithms/Routes/CompleteRouteBuilder.cs [hunk 3]
```diff
@@ -470,36 +470,36 @@ private void AppendRoute(Route route, IAttributeCollection shortcutMeta)
...
             path.AddToListAsVertices(pathList);
-            return CompleteRouteBuilder.TryBuild(db, profile, source, target, pathList);
+            return CompleteRouteBuilder.TryBuild(db, profile, source, target, pathList, cancellationToken);
         }
 
         /// <summary>
         /// Builds a route.
         /// </summary>
-        public static Route Build(RouterDb db, Profile profile, RouterPoint source, RouterPoint target, List<uint> path)
+        public static Route Build(RouterDb db, Profile profile, RouterPoint source, RouterPoint target, List<uint> path, CancellationToken cancellationToken)
```
Comment (airbreather) [inline src/Itinero/Algorithms/Routes/CompleteRouteBuilder.cs]: Missed defaulting here?

File: src/Itinero/Algorithms/Routes/CompleteRouteBuilder.cs [hunk 4]
```diff
@@ -470,36 +470,36 @@ private void AppendRoute(Route route, IAttributeCollection shortcutMeta)
...
         {
-            return CompleteRouteBuilder.TryBuild(db, profile, source, target, path).Value;
+            return CompleteRouteBuilder.TryBuild(db, profile, source, target, path, cancellationToken).Value;
         }
 
         /// <summary>
         /// Builds a route.
         /// </summary>
-        public static Result<Route> TryBuild(RouterDb db, Profile profile, RouterPoint source, RouterPoint target, List<uint> path)
+        public static Result<Route> TryBuild(RouterDb db, Profile profile, RouterPoint source, RouterPoint target, List<uint> path, CancellationToken cancellationToken)
```
Comment (airbreather) [inline src/Itinero/Algorithms/Routes/CompleteRouteBuilder.cs]: Missed defaulting here?

File: src/Itinero/Algorithms/Routes/FastRouteBuilder.cs [hunk 5]
```diff
@@ -391,37 +392,40 @@ private Speed GetSpeedFor(ushort profileId)
         /// <summary>
         /// Builds a route.
         /// </summary>
-        public static Route Build(RouterDb db, Profile profile, Func<ushort, Profiles.Factor> getFactor, RouterPoint source, RouterPoint target, EdgePath<float> path)
+        public static Route Build(RouterDb db, Profile profile, Func<ushort, Profiles.Factor> getFactor, RouterPoint source, RouterPoint target, EdgePath<float> path,
+            CancellationToken cancellationToken)
```
Comment (airbreather) [inline src/Itinero/Algorithms/Routes/FastRouteBuilder.cs]: Missed defaulting here?

File: src/Itinero/Algorithms/Routes/FastRouteBuilder.cs [hunk 6]
```diff
@@ -391,37 +392,40 @@ private Speed GetSpeedFor(ushort profileId)
...
-            return FastRouteBuilder.TryBuild(db, profile, getFactor, source, target, path).Value;
+            return FastRouteBuilder.TryBuild(db, profile, getFactor, source, target, path, cancellationToken).Value;
         }
 
         /// <summary>
         /// Builds a route.
         /// </summary>
-        public static Result<Route> TryBuild(RouterDb db, Profile profile, Func<ushort, Profiles.Factor> getFactor, RouterPoint source, RouterPoint target, EdgePath<float> path)
+        public static Result<Route> TryBuild(RouterDb db, Profile profile, Func<ushort, Profiles.Factor> getFactor, RouterPoint source, RouterPoint target, EdgePath<float> path,
+            CancellationToken cancellationToken)
```
Comment (airbreather) [inline src/Itinero/Algorithms/Routes/FastRouteBuilder.cs]: Missed defaulting here?

File: src/Itinero/Algorithms/Routes/FastRouteBuilder.cs [hunk 7]
```diff
@@ -391,37 +392,40 @@ private Speed GetSpeedFor(ushort profileId)
...
-            return FastRouteBuilder.TryBuild(db, profile, getFactor, source, target, pathList);
+            return FastRouteBuilder.TryBuild(db, profile, getFactor, source, target, pathList, cancellationToken);
         }
 
         /// <summary>
         /// Builds a route.
         /// </summary>
-        public static Route Build(RouterDb db, Profile profile, Func<ushort, Profiles.Factor> getFactor, RouterPoint source, RouterPoint target, List<uint> path)
+        public static Route Build(RouterDb db, Profile profile, Func<ushort, Profiles.Factor> getFactor, RouterPoint source, RouterPoint target, List<uint> path,
+            CancellationToken cancellationToken)
```
Comment (airbreather) [inline src/Itinero/Algorithms/Routes/FastRouteBuilder.cs]: Missed defaulting here?

File: src/Itinero/Algorithms/Routes/FastRouteBuilder.cs [hunk 8]
```diff
@@ -391,37 +392,40 @@ private Speed GetSpeedFor(ushort profileId)
...
-            return FastRouteBuilder.TryBuild(db, profile, getFactor, source, target, path).Value;
+            return FastRouteBuilder.TryBuild(db, profile, getFactor, source, target, path, cancellationToken).Value;
         }
 
         /// <summary>
         /// Builds a route.
         /// </summary>
         public static Result<Route> TryBuild(RouterDb db, Profile profile, Func<ushort, Profiles.Factor> getFactor, 
-            RouterPoint source, RouterPoint target, List<uint> path)
+            RouterPoint source, RouterPoint target, List<uint> path, CancellationToken cancellationToken)
```
Comment (airbreather) [inline src/Itinero/Algorithms/Routes/FastRouteBuilder.cs]: Missed defaulting here?

File: src/Itinero/Algorithms/Search/Hilbert/HilbertExtensions.cs [hunk 9]
```diff
@@ -634,7 +635,8 @@ public static long Distance(this RoutingNetwork graph, int n, uint vertex)
         /// </summary>
         /// <returns></returns>
         public static uint SearchClosestEdge(this GeometricGraph graph, float latitude, float longitude,
-            float latitudeOffset, float longitudeOffset, float maxDistanceMeter, Func<GeometricEdge, bool> isOk)
+            float latitudeOffset, float longitudeOffset, float maxDistanceMeter, Func<GeometricEdge, bool> isOk,
+            CancellationToken cancellationToken)
```
Comment (airbreather) [inline src/Itinero/Algorithms/Search/Hilbert/HilbertExtensions.cs]: Missed defaulting here?

File: src/Itinero/Algorithms/Shortcuts/RouterDbExtensions.cs [hunk 10]
```diff
@@ -30,7 +31,7 @@ public static class RouterDbExtensions
         /// <summary>
         /// Adds multiple shortcut db's at the same time.
         /// </summary>
-        public static void AddShortcuts(this RouterDb routerDb, params ShortcutSpecs[] shortcutSpecs)
+        public static void AddShortcuts(this RouterDb routerDb, CancellationToken cancellationToken = new CancellationToken(), params ShortcutSpecs[] shortcutSpecs)
```
Comment (airbreather) [inline src/Itinero/Algorithms/Shortcuts/RouterDbExtensions.cs]: I'm not in love with how this makes the callers look... maybe this is one of those cases where it makes sense to add an overload with the cancellation token at the cost of the cancellation-aware overload's `shortcutSpecs` not being `params`?

**Comment by JoeCooper** [comment]
Hey, thanks for having a look through this. At a glance, it looks like you’re right on all points. Furthermore if you or anyone claims to use a given method from outside Itinero, than unless a default parameter is provided, it is a breaking change and therefore incorrect. When I’m in the office on Tuesday I’ll put in a new commit.

File: src/Itinero/Algorithms/Contracted/Dual/RouterExtensions.cs [hunk 11]
```diff
@@ -32,7 +33,7 @@ internal static class RouterExtensions
         /// Calculates a many-to-many weight matrix using a dual edge-based graph.
         /// </summary>
         internal static Result<T[][]> CalculateManyToMany<T>(this ContractedDb contractedDb, RouterDb routerDb, Profile profile, WeightHandler<T> weightHandler, 
-            RouterPoint[] sources, RouterPoint[] targets, T max) where T : struct
+            RouterPoint[] sources, RouterPoint[] targets, T max, CancellationToken cancellationToken = new CancellationToken()) where T : struct
```
[the remark this answers is not in the document]
Comment (JoeCooper) [reply src/Itinero/Algorithms/Contracted/Dual/RouterExtensions.cs]: @xivk can we use this C# 7.1 feature?

**Comment by JoeCooper** [comment]
@airbreather I've changed a bit based on your comments, although I'm putting off using default until @xivk confirms we can use C# 7.1 in this project. What do you think about the interface in Shortcuts/RouterDbExtensions now?

**Comment by airbreather** [comment]
> What do you think about the interface in Shortcuts/RouterDbExtensions now?

Hmm, this works just fine for me:
```csharp
        /// <summary>
        /// Adds multiple shortcut db's at the same time.
        /// </summary>
        public static void AddShortcuts(this RouterDb routerDb, params ShortcutSpecs[] shortcutSpecs) => AddShortcuts(routerDb, shortcutSpecs, new CancellationToken());

        /// <summary>
        /// Adds multiple shortcut db's at the same time.
        /// </summary>
        public static void AddShortcuts(this RouterDb routerDb, ShortcutSpecs[] shortcutSpecs, CancellationToken cancellationToken = new CancellationToken())
        {
```

**Comment by JoeCooper** [comment]
That's nice. I didn't know you could do that. I'll stick that in there now.

**Comment by airbreather** [comment]
That looks good now.

One thing to point out is that now the only two files under the `test` folder that need to change are:
- `test\Itinero.Test\RouterMock.cs`
- `test\Itinero.Test\Algorithms\Search\MockResolver.cs`

It appears that all other changes under `test\` could be reverted... I determined this by doing a mass find/replace of `Run(new CancellationToken())` to `Run()`, then comparing the `test\` folder to what it was before, and observing that the only remaining changes outside those two files were in the `using` blocks.

**Comment by JoeCooper** [comment]
I'll evaluate that now. Thanks for the heads up.

**Comment by JoeCooper** [comment]
I've gone ahead and reverted the changes to the unit tests.

**Comment by JoeCooper** [comment]
Looks like 'default' is already in use, in Algorithms/Collections/BinaryHeap.cs line 191. I'm going to go ahead and use it.

**Comment by airbreather** [comment]
If you merge JoeCooper/routing#1, then the number of files touched in this PR should go back down, such that after this PR gets merged, all those files don't show this as their last change...

I did a little trick to revert all of them at once:
1. Check out original version
2. Copy `test` folder off
3. Check out your branch
4. Replace the `test` folder with the copy
5. Undo changes to the two I identified above
6. Commit and push to my fork

**Comment by airbreather** [comment]
This looks fine to me now :+1:.

The C# 7.1 `default` vs. `new CancellationToken()` thing would be nice to have, but I don't care and it could happen later.

**Comment by JoeCooper** [comment]
It does tell me explicitly that the project is not configured to use 7.1, but if @xivk allows it than we're good to go on that point too.

**Comment by JoeCooper** [comment]
Oh, you can approve the PR? I didn't realize you had that relationship to the project. Can you approve a move to 7.1?

**Comment by airbreather** [comment]
Approval just means I reviewed it and I'm OK with it.  I did leave some commentary on #156 endorsing this approach, so I thought it only appropriate that I give feedback on the implementation.

**Comment by JoeCooper** [comment]
All clear.

**Comment by airbreather** [comment]
For example, one thing I mentioned during my review was overloads vs. default parameters... I'm OK with this myself, but that's only because I don't have any backwards compatibility concerns (my company still isn't using Itinero yet)... @xivk may have other opinions, heh.

**Comment by JoeCooper** [comment]
Yes, that makes sense. I brought them up in the issue thread because default parameters break the ABI, but Ben decided this was acceptable.

**Comment by xivk** [comment]
I'm on holiday this week, sorry... I'll have a look at all this next week.

Thanks for the work done on this already!

**Comment by JoeCooper** [comment]
I was out for a week. For various reasons I might not have an opportunity to add more cancellation tests, but if the interfaces are updated, others can add them on their own critical paths on an as needed basis and the feature will accumulate over time.

File: src/Itinero/Algorithms/Contracted/Dual/RouterExtensions.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (xivk) [reply src/Itinero/Algorithms/Contracted/Dual/RouterExtensions.cs]: I think so, any downside to using C# 7.1 features at this point? I guess it stops people from using older VS versions?

File: src/Itinero/Algorithms/IAlgorithm.cs [hunk 12]
```diff
@@ -42,7 +44,7 @@ bool HasSucceeded
         /// <summary>
         /// Runs the algorithm.
         /// </summary>
-        void Run();
+        void Run(CancellationToken cancellationToken);
```
[the remark this answers is not in the document]
Comment (xivk) [reply src/Itinero/Algorithms/IAlgorithm.cs]: I agree, if we can do this, we should.

**Comment by xivk** [comment]
OK I left some obsolete comments above, I should have read everything first. 

This looks really good, it has minimal impact and adds a valuable feature. Thanks a lot!

About the C# 7.1, let's go for it.

About the binary compat issue, what would it take to guarantee this? I guess replacing defaults with overloads everywhere? Most usecases I know of don't upgrade without rebuilding so I don't think it's a big issue but perhaps something to aim for in the future? How do other (well-managed) .NET packages handle this?

**Comment by airbreather** [comment]
>>@xivk can we use this C# 7.1 feature?
>
>I think so, any downside to using C# 7.1 features at this point? I guess it stops people from using older VS versions?

Yeah, using C# 7.1 features pretty much just makes it so it won't build in VS2017 15.2 and below.  You also have to enable it specifically by putting `<LangVersion>X</LangVersion>` into a `PropertyGroup` in the `.csproj` file where you want to use the stuff, where X is one of:
1. `latest` (assuming you're on a new enough VS2017 version)
1. `7.1` (VS2017 15.3 and above)
1. `7.2` (VS2017 15.5 and above)
1. `7.3` (VS2017 15.7 and above, still in preview at the time of writing)

At work, I've advised that we use an explicit version number instead of `latest`, because it kills builds much earlier and offers a more helpful message if the compiler doesn't support that version instead of just a generic syntax error.  YMMV.

> About the binary compat issue, what would it take to guarantee this? I guess replacing defaults with overloads everywhere? Most usecases I know of don't upgrade without rebuilding so I don't think it's a big issue but perhaps something to aim for in the future? How do other (well-managed) .NET packages handle this?

As far as I know, yes, adding overloads is the way to go for this kind of change to ensure both binary and source compatibility.  Adding optional parameters feels like it can be a pretty hard breaking change since it can cascade downstream: assuming perfect NuGet version specs, then in the best case, any packages depending on Itinero that have their own downstream dependents would have to mark their own packages as depending on "pre-`CancellationToken`" (maximum) or "post-`CancellationToken`" (higher minimum) versions of Itinero to ensure that applications can't break things simply by bumping up their own version of Itinero, so the default-parameter approach is technically not semver-compliant if this happens before a major version bump (1.x to 2.x).

**Comment by airbreather** [comment]
> The correct way to handle this must be overloads, then. When I did this, that did not occur to me.
>
> I will find some time this week to make another commit.

I can pick this up... I don't think there's a sane way to avoid breaking external stuff that adds their own subclasses of `AlgorithmBase`, but everything else should be easy enough (just tedious, which I'm fine with, since I have some time to kill).

**Comment by airbreather** [comment]
Hmm, right, that brings up an interesting problem: some of the static methods have their own optional parameters other than `CancellationToken`, e.g.:

```csharp
public static void SplitLongEdges(this RouterDb db, Action<uint> newVertex = null, CancellationToken cancellationToken = new CancellationToken())
```

It can't tack on the `CancellationToken` at the end while still keeping optional values for those other parameters.  So those other optional parameters would become required when calling the method with a `CancellationToken` overload.

I'm going to go with having callers give explicit values when using a `CancellationToken` overload for 1.x, with the idea that 2.x can make this specific thing nicer.

**Comment by airbreather** [comment]
My branch [`cancellation-overloads`](https://github.com/airbreather/routing/tree/cancellation-overloads) is more-or-less the size of the approach that uses... I still need to do a bit of self-review on it.

That feels like too much code just to ensure binary compatibility, though.

**Comment by xivk** [comment]
> That feels like too much code just to ensure binary compatibility, though.

I feel like that all the time when adding new features, the technical debt keeps growing. That's why a v2 is long overdue.

That being said the [`cancellation-overloads`](https://github.com/airbreather/routing/tree/cancellation-overloads) branch looks good, if we can bring that in then this is done.

**Comment by airbreather** [comment]
I've made a few updates and created JoeCooper/routing#2... once @JoeCooper merges that into his branch, this PR should automatically update to have the overloads instead of the optional parameters.

**Comment by JoeCooper** [comment]
I've merged in the updates from @airbreather. Sorry it took a while. April has been an absolute clusterfrack. I'm back in the office though.

**Comment by JoeCooper** [comment]
Merged again.

**Comment by airbreather** [comment]
:+1: this looks good to me now.

**Comment by xivk** [comment]
Thanks @JoeCooper @airbreather :+1:

**Comment by JoeCooper** [comment]
Thanks everyone

**Pull request merged**